File: semantic/src/iot/semantic/expr.clj
(ns iot.semantic.expr
  "IoT 자동화 레시피의 의미를 표현하는 중간 표현(IR).

  핵심 통찰: Clojure에서 Expr은 '타입'이 아니라 '맵'이다.
  Go에서 struct + Op enum + Validate()로 100줄 걸리던 것이
  여기서는 그냥 데이터다.

  {:op :eq
   :children [{:op :state-ref :device \"sensor\" :attr \"motion\"}
              {:op :lit :value true}]}

  이것이 전부다. 괄호가 이미 AST이고, 맵이 이미 타입이다.
  Sussman 교수님이 Scheme으로 가르친 것의 핵심:
  코드와 데이터의 구분이 사라지면, 변환은 자연스러워진다.

  설계 원칙:
  1. 데이터 우선 — defrecord/deftype 대신 순수한 맵
  2. 열린 확장 — :op은 키워드. 누구든 :ha/choose, :tuya/precondition 추가 가능
  3. Spec으로 검증 — 구조 검증은 clojure.spec이 담당
  4. 조합자(combinator) — 함수가 Expr을 만들고, 함수가 Expr을 변환한다")

;; ─── 리프 생성자 ─────────────────────────────────────────────

(defn lit
  "리터럴 값. 숫자, 문자열, 불리언.
  Go에서 Lit(v any)가 타입을 잃어버리는 문제가 여기서는 없다.
  Clojure의 값은 자기 타입을 알고 있다."
  [v]
  {:op :lit :value v})

(defn state-ref
  "디바이스 상태 참조.
  플랫폼마다 ID 체계가 다르지만, 구조는 같다:
  '어떤 장치의 어떤 속성'."
  ([device attr]
   {:op :state-ref :device device :attr attr})
  ([device component attr]
   {:op :state-ref :device device :component component :attr attr}))

(defn time-ref
  "시간 참조. 'now', '22:00', 'sunrise+30m'."
  [v]
  {:op :time-ref :value v})

;; ─── 비교 조합자 ─────────────────────────────────────────────

(defn eq      [a b] {:op :eq      :children [a b]})
(defn ne      [a b] {:op :ne      :children [a b]})
(defn gt      [a b] {:op :gt      :children [a b]})
(defn ge      [a b] {:op :ge      :children [a b]})
(defn lt      [a b] {:op :lt      :children [a b]})
(defn le      [a b] {:op :le      :children [a b]})
(defn between [v lo hi] {:op :between :children [v lo hi]})

(defn in-set
  "멤버십 검사. val ∈ {a, b, c}."
  [val & members]
  {:op :in :children (into [val] members)})

(defn contains-str [a b] {:op :contains :children [a b]})

;; ─── 논리 조합자 ─────────────────────────────────────────────

(defn and-expr
  "모든 조건이 참. Go의 AndExpr(...*Expr)과 같지만,
  가변 인자가 자연스럽다."
  [& children]
  {:op :and :children (vec children)})

(defn or-expr  [& children] {:op :or  :children (vec children)})
(defn not-expr [child]      {:op :not :children [child]})

;; ─── 시퀀스 조합자 ─────────────────────────────────────────────

(defn seq-expr
  "순차 실행. 순서가 의미를 가진다.
  (seq-expr (cmd ...) (delay 5) (cmd ...))"
  [& children]
  {:op :seq :children (vec children)})

(defn parallel [& children] {:op :parallel :children (vec children)})

;; ─── 액션 생성자 ─────────────────────────────────────────────

(defn cmd
  "디바이스 명령."
  [device attr value]
  {:op :command :device device :attr attr :value value})

(defn delay-expr
  "지연(초 단위)."
  [seconds]
  {:op :delay :value seconds})

(defn notify
  "알림 메시지."
  [message]
  {:op :notify :value message})

(defn scene
  "씬 활성화."
  [scene-id]
  {:op :scene :value scene-id})

;; ─── 메타데이터 ─────────────────────────────────────────────

(defn with-meta*
  "플랫폼 특수 메타데이터 부착.
  Clojure의 metadata와 구분하기 위해 :meta 키 사용.
  org-mode의 #+ATTR_LATEX:과 같은 역할."
  [expr k v]
  (assoc-in expr [:meta k] v))

;; ─── 레시피: trigger + condition + action의 묶음 ────────────

(defn recipe
  "하나의 자동화 규칙.
  IoT recipe는 결국 이 세 가지의 조합이다:
  - trigger:   무엇이 일어나면
  - condition: 어떤 조건 아래서
  - actions:   무엇을 한다"
  [& {:keys [id name trigger condition actions]}]
  (cond-> {:id id :name name :actions actions}
    trigger   (assoc :trigger trigger)
    condition (assoc :condition condition)))

;; ─── 트리 순회 유틸 ─────────────────────────────────────────

(defn walk-expr
  "Expr 트리의 모든 노드에 f를 적용한다 (bottom-up).
  Lisp에서는 이것이 내장이다. Go에서는 매번 재귀를 짰다.

  (walk-expr identity expr)           ; 항등 변환
  (walk-expr #(assoc % :visited true) expr)  ; 모든 노드에 표시"
  [f expr]
  (if-let [children (:children expr)]
    (f (assoc expr :children (mapv (partial walk-expr f) children)))
    (f expr)))

(defn fold-expr
  "Expr 트리를 하나의 값으로 접는다 (bottom-up).
  노드 수 세기, 디바이스 참조 수집 등.

  (fold-expr (fn [acc node] (inc acc)) 0 expr)  ; 노드 수"
  [f init expr]
  (let [child-result (reduce (fn [acc child]
                               (fold-expr f acc child))
                             init
                             (:children expr []))]
    (f child-result expr)))

(defn device-refs
  "트리에서 참조된 모든 디바이스 ID를 수집한다."
  [expr]
  (fold-expr
   (fn [acc node]
     (if-let [d (:device node)]
       (conj acc d)
       acc))
   #{}
   expr))

(defn depth
  "트리의 최대 깊이."
  [expr]
  (if-let [children (seq (:children expr))]
    (inc (apply max (map depth children)))
    1))

(defn node-count
  "트리의 총 노드 수."
  [expr]
  (fold-expr (fn [acc _] (inc acc)) 0 expr))

File: archive/go/core/expr/expr.go
// Package expr defines a structured expression tree for IoT automation rules.
//
// This replaces the unstructured map[string]any Config with a typed AST
// that preserves semantic meaning across platform conversions.
//
// Design philosophy: Sussman's SDF (Software Design for Flexibility) —
// composable parts with standardized interfaces. Each Expr node is a
// combinator that can be freely composed.
//
// The key insight: flexibility that preserves structure IS durability.
// A well-typed expression tree is both flexible (any platform maps to it)
// and durable (meaning survives transformation).
//
//	Tuya:  dp_id=1, comparator="==", value=true
//	HA:    entity_id=switch.lamp, to="on"
//	Both → Expr{Op: Eq, Left: StateRef("switch.lamp"), Right: Literal("on")}
package expr

import "encoding/json"

// Op is an expression operator.
type Op string

const (
	// ── Comparison operators ──

	// Eq checks equality: state == value.
	Eq Op = "eq"
	// Ne checks inequality: state != value.
	Ne Op = "ne"
	// Gt checks greater than: state > value.
	Gt Op = "gt"
	// Ge checks greater than or equal: state >= value.
	Ge Op = "ge"
	// Lt checks less than: state < value.
	Lt Op = "lt"
	// Le checks less than or equal: state <= value.
	Le Op = "le"
	// Between checks range: low <= state <= high.
	Between Op = "between"
	// In checks membership: state in {a, b, c}.
	In Op = "in"
	// Contains checks substring: state contains "x".
	Contains Op = "contains"

	// ── Logical combinators ──

	// And requires all children to be true.
	And Op = "and"
	// Or requires any child to be true.
	Or Op = "or"
	// Not negates the child expression.
	Not Op = "not"

	// ── Sequence combinators ──

	// Seq executes children in order. [action1, delay, action2]
	Seq Op = "seq"
	// Parallel executes children concurrently.
	Parallel Op = "parallel"

	// ── Leaf nodes ──

	// Literal is a concrete value: "on", 25, true.
	Literal Op = "literal"
	// StateRef reads a device attribute: device.temperature.
	StateRef Op = "state_ref"
	// TimeRef references a time value: "22:00", sunrise+offset.
	TimeRef Op = "time_ref"

	// ── Action nodes ──

	// Command issues a device command: turn_on, set_temperature(22).
	Command Op = "command"
	// Delay waits for a duration.
	Delay Op = "delay"
	// Notify sends a notification.
	Notify Op = "notify"
	// Scene activates a scene or runs another automation.
	Scene Op = "scene"
)

// Expr is a node in the expression tree.
// It is the universal intermediate representation (IR) for IoT automation rules.
//
// Composition examples:
//
//	temperature > 25:
//	  {Op: Gt, Children: [{Op: StateRef, Ref: {ID: "sensor.temp"}}, {Op: Literal, Value: 25}]}
//
//	time between 22:00-06:00:
//	  {Op: Between, Children: [{Op: TimeRef, Value: "now"}, {Op: Literal, Value: "22:00"}, {Op: Literal, Value: "06:00"}]}
//
//	(temp > 25) AND (light == "off"):
//	  {Op: And, Children: [<temp>25>, <light==off>]}
type Expr struct {
	Op       Op             `json:"op"`
	Children []*Expr        `json:"children,omitempty"`
	Ref      *DeviceRef     `json:"ref,omitempty"`
	Value    any            `json:"value,omitempty"`
	Meta     map[string]any `json:"meta,omitempty"`
}

// DeviceRef identifies a device and its attribute across platforms.
//
// The key problem: Tuya uses "dp_id=1" on device "xxx",
// HA uses "entity_id=sensor.temperature", SmartThings uses
// "device.main.temperatureMeasurement.temperature".
// DeviceRef normalizes these into a single structure.
type DeviceRef struct {
	// DeviceID is the platform-native device identifier.
	DeviceID string `json:"device_id"`
	// Attribute is the property/capability being referenced.
	// HA: state attribute (e.g., "temperature", "state")
	// Tuya: dp_id (e.g., "dp_1", "dp_4")
	// SmartThings: capability.attribute (e.g., "switch.switch")
	// Google: trait.attribute (e.g., "OnOff.on")
	// Homey: capability (e.g., "onoff", "measure_temperature")
	Attribute string `json:"attribute,omitempty"`
	// Component for multi-component devices (ST: "main", HA: domain).
	Component string `json:"component,omitempty"`
}

// ── Constructors: SDF-style combinators ──

// Lit creates a literal value node.
func Lit(v any) *Expr { return &Expr{Op: Literal, Value: v} }

// State creates a device state reference.
func State(deviceID, attribute string) *Expr {
	return &Expr{Op: StateRef, Ref: &DeviceRef{DeviceID: deviceID, Attribute: attribute}}
}

// StateWithComponent creates a device state ref with a component.
func StateWithComponent(deviceID, component, attribute string) *Expr {
	return &Expr{Op: StateRef, Ref: &DeviceRef{
		DeviceID: deviceID, Attribute: attribute, Component: component,
	}}
}

// Time creates a time reference.
func Time(value string) *Expr { return &Expr{Op: TimeRef, Value: value} }

// ── Comparison combinators ──

// cmp creates a binary comparison expression.
func cmp(op Op, left, right *Expr) *Expr {
	return &Expr{Op: op, Children: []*Expr{left, right}}
}

// EqExpr creates an equality check: left == right.
func EqExpr(left, right *Expr) *Expr { return cmp(Eq, left, right) }

// NeExpr creates an inequality check: left != right.
func NeExpr(left, right *Expr) *Expr { return cmp(Ne, left, right) }

// GtExpr creates a greater-than check: left > right.
func GtExpr(left, right *Expr) *Expr { return cmp(Gt, left, right) }

// GeExpr creates a greater-or-equal check: left >= right.
func GeExpr(left, right *Expr) *Expr { return cmp(Ge, left, right) }

// LtExpr creates a less-than check: left < right.
func LtExpr(left, right *Expr) *Expr { return cmp(Lt, left, right) }

// LeExpr creates a less-or-equal check: left <= right.
func LeExpr(left, right *Expr) *Expr { return cmp(Le, left, right) }

// BetweenExpr creates a range check: low <= val <= high.
func BetweenExpr(val, low, high *Expr) *Expr {
	return &Expr{Op: Between, Children: []*Expr{val, low, high}}
}

// InExpr creates a membership check: val in set.
func InExpr(val *Expr, set ...*Expr) *Expr {
	children := make([]*Expr, 0, 1+len(set))
	children = append(children, val)
	children = append(children, set...)
	return &Expr{Op: In, Children: children}
}

// ── Logical combinators ──

// AndExpr combines conditions: all must be true.
func AndExpr(children ...*Expr) *Expr { return &Expr{Op: And, Children: children} }

// OrExpr combines conditions: any must be true.
func OrExpr(children ...*Expr) *Expr { return &Expr{Op: Or, Children: children} }

// NotExpr negates a condition.
func NotExpr(child *Expr) *Expr { return &Expr{Op: Not, Children: []*Expr{child}} }

// ── Sequence combinators ──

// SeqExpr creates a sequential execution: [a, then b, then c].
func SeqExpr(children ...*Expr) *Expr { return &Expr{Op: Seq, Children: children} }

// ParallelExpr creates parallel execution: [a and b and c simultaneously].
func ParallelExpr(children ...*Expr) *Expr { return &Expr{Op: Parallel, Children: children} }

// ── Action constructors ──

// Cmd creates a device command expression.
func Cmd(deviceID, attribute string, value any) *Expr {
	return &Expr{
		Op:    Command,
		Ref:   &DeviceRef{DeviceID: deviceID, Attribute: attribute},
		Value: value,
	}
}

// DelayExpr creates a delay action.
func DelayExpr(seconds float64) *Expr {
	return &Expr{Op: Delay, Value: seconds}
}

// NotifyExpr creates a notification action.
func NotifyExpr(message string) *Expr {
	return &Expr{Op: Notify, Value: message}
}

// SceneExpr creates a scene activation action.
func SceneExpr(sceneID string) *Expr {
	return &Expr{Op: Scene, Value: sceneID}
}

// ── Helpers ──

// WithMeta attaches platform-specific metadata to an expression.
// Returns the same Expr for chaining: Cmd(...).WithMeta("service", "light.turn_on")
func (e *Expr) WithMeta(key string, value any) *Expr {
	if e.Meta == nil {
		e.Meta = make(map[string]any)
	}
	e.Meta[key] = value
	return e
}

// IsLeaf returns true if this is a leaf node (no children).
func (e *Expr) IsLeaf() bool {
	return len(e.Children) == 0
}

// IsComparison returns true if this is a comparison operator.
func (e *Expr) IsComparison() bool {
	switch e.Op {
	case Eq, Ne, Gt, Ge, Lt, Le, Between, In, Contains:
		return true
	}
	return false
}

// IsCombinator returns true if this is a logical/sequence combinator.
func (e *Expr) IsCombinator() bool {
	switch e.Op {
	case And, Or, Not, Seq, Parallel:
		return true
	}
	return false
}

// IsAction returns true if this is an action node.
func (e *Expr) IsAction() bool {
	switch e.Op {
	case Command, Delay, Notify, Scene:
		return true
	}
	return false
}

// String returns a human-readable representation.
func (e *Expr) String() string {
	if e == nil {
		return "<nil>"
	}
	b, _ := json.Marshal(e)
	return string(b)
}
